### DISCUSSION TITLE
How to version Liquid Prompt themes

### DISCUSSION THREAD

**Discussion by Rycieos** [liquidprompt/liquidprompt#799] [Ideas]
### Context

@Rycieos:
> Let me know if you want me to wait to release v2.2 here... I don't think that is necessary; we could release v1.0 [in the liquidprompt-powerline repo].

@nojhan:
> I also thought that it will be probably easier to understand for users if the themes follows the same versions than LP (i.e. being versioned 2.2.0 when they support LP 2.2.0). That way, you immediately know if the versions are aligned.

## Problem statement
Versioning software that depends closely on other software is hard. The Liquid Prompt project is going to have a number of themes that are owned by the project, and determining how to version them is an important step. If we can agree on a good method, hopefully other community owned themes can follow the practice.

## Options
1. Keep theme versions matching Liquid Prompt versions. This will make it very obvious when a theme will work with Liquid Prompt. The downside is new features or bug fixes can't be added without version drift happening.
2. Use a dual versioning system. Something like `v2.2.0-1.0.0` for example. The first version being the Liquid Prompt version, and the second being the theme version. This could be confusing for users, and still locks each theme version to being compatible with exactly one Liquid Prompt version.
3. Force no relationship between versions, but have a compatibility matrix ([example](https://projectcontour.io/resources/compatibility-matrix/)) that defines what versions work with what versions of Liquid Prompt.

My plan was to do 3), but I am open for feedback.

I really would like to see themes supporting multiple Liquid Prompt versions at once, but that might not be easy or even worth it.

I really do think that 1) is not a good idea, since there would be no way to fix bugs in a theme.

**Comment by nojhan** [comment]
1) If we stick to semantic versioning, then MAJOR.MINOR is sufficient to test for compatibility, and PATCH can vary between LP and themes. That way, bug fixes are possible in themes or LP without hindering compatibility tests and readability of the versions by the user.
There may still be some issues when the themes do a backward-incompatible change against itself (like renaming a config variable), but I think it's a minor issue worth the effort.
3) This is easier on the devs, but complex for the users. At the same time, people are probably just cloning whatever is the last stable version of LP and theme, and just don't bother with versions anyway.

**Reply by Rycieos** [reply]
> people are probably just cloning whatever is the last stable version of LP and theme, and just don't bother with versions anyway.

This is likely true, and means this probably isn't as important as I originally thought.

> If we stick to semantic versioning, then MAJOR.MINOR is sufficient to test for compatibility, and PATCH can vary between LP and themes. That way, bug fixes are possible in themes or LP without hindering compatibility tests and readability of the versions by the user.
There may still be some issues when the themes do a backward-incompatible change against itself (like renaming a config variable), but I think it's a minor issue worth the effort.

This is probably all fair. But I would argue then for a 4th option: for no versioning at all.

Since most will track the latest anyway, why bother versioning. Simply mention which Liquid Prompt version it currently supports.

**Reply by nojhan** [reply]
That's actually what I did with the version guards, indeed.

Now I am also wondering what option people maintaining packages here and
there would prefer. I don't maintain any package, so I have no clue, but
guessing version matters a bit for them? Or are commit hashes enough?

**Reply by nojhan** [reply]
I have been exposed to this problem in another setting, and the depending software was using an additional specific tag, decoupled from its own version scheme: just a tag prefixed by the dependency name.

That would be, two simultaneous tags for each version:

- a regular version: e.g. `v1.2.3`
-  a prefixed version indicating the dependency, e.g. `liquiprompt-v2.2`

It was actually easy to use, since I was able to spot the necessary version and checkout rapidly.

**Reply by Rycieos** [reply]
I like that. How would those versions be stored and displayed? Usually, with only one version, it can be the git tag name, and a simple string that is printed in help text and such.

**Reply by nojhan** [reply]
My themes have a version guard, with an hard-coded required LP version (e.g. [in DotMatrix](https://github.com/nojhan/lp-dotmatrix/blob/74086c9fdd35edac376d5e03585ff80994cd237c/dotmatrix.theme#L8)). Maybe LP can provide this as a function, so that theme would just have to call it with their supported version.

Using the git tag within the theme code would require a build step, which does not seems practical.

**Reply by Rycieos** [reply]
Don't we already have that with [`_lp_version_greatereq()`](https://liquidprompt.readthedocs.io/en/v2.2.0-rc.2/functions/util.html#lp_version_greatereq)?

**Reply by nojhan** [reply]
I was thinking of something more high level, like to have the theme set a prefixed variable `__NAME_theme_required_version` and let LP check when it loads the theme, and warn the user in case of mismatch.

**Reply by Rycieos** [reply]
Wow, that's such a good idea actually; it has grown on me all night. I'll bang out a PR for that quick.
